### PR TITLE
Android Release v7.4.0 

### PR DESCRIPTION
**Pull request by ywang-nr**
[NR-239869](https://new-relic.atlassian.net/browse/NR-239869) Mobile Logging
[NR-247772](https://new-relic.atlassian.net/browse/NR-247772) Application Exit Info reporting
[NR-169834](https://new-relic.atlassian.net/browse/NR-169834) Background Monitoring
[NR-266913](https://new-relic.atlassian.net/browse/NR-266913) setUserID()
[NR-59017](https://new-relic.atlassian.net/browse/NR-59017) TicToc time calculation
[NR-275931](https://new-relic.atlassian.net/browse/NR-275931) Resolve ANRs in eventStore
[NR-256169](https://new-relic.atlassian.net/browse/NR-256169) Resolve exceptionMessage exceed 4096 characters in mobilehandledexception

[NR-239869]: https://new-relic.atlassian.net/browse/NR-239869?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-247772]: https://new-relic.atlassian.net/browse/NR-247772?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-169834]: https://new-relic.atlassian.net/browse/NR-169834?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[NR-266913]: https://new-relic.atlassian.net/browse/NR-266913?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[NR-59017]: https://new-relic.atlassian.net/browse/NR-59017?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-275931]: https://new-relic.atlassian.net/browse/NR-275931?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-256169]: https://new-relic.atlassian.net/browse/NR-256169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ywang-nr** [comment]
@cthomas-newrelic Can you help to take a look at this test, it's failing here? Thank you.

com.newrelic.agent.android.NewRelicTest > testLogReportingFeature FAILED
    java.lang.AssertionError at NewRelicTest.java:1088

File: agent-core/build.gradle [hunk 1]
```diff
@@ -38,6 +38,7 @@ jar {
 
 dependencies {
     compileOnly fileTree(dir: 'libs', include: '*.jar')
+    compileOnly "org.jetbrains:annotations:15.0"
```
Comment (cthomas-newrelic) [inline agent-core/build.gradle:41]: Inspect: may be unnecessary

File: agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+        LogReportState(final String extension) {
+            this.extension = extension;
+        }
+
+        public String asExtension() {
+            return String.format(Locale.getDefault(), ".%s", extension);
+        }
+    }
+
+    static int VORTEX_PAYLOAD_LIMIT = (1024 * 1023);                // Vortex upload limit: ~1 MB (10^6 B) compressed (less some padding)
```
Comment (ywang-nr) [inline agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java:71]: change size to 1000

File: agent-core/src/main/java/com/newrelic/agent/android/payload/PayloadSender.java [hunk 3]
```diff
@@ -107,14 +111,15 @@ protected void onFailedUpload(String errorMsg) {
     @SuppressWarnings("NewApi")
     public PayloadSender call() throws Exception {
         try {
-            timer.tic();
+            byte[] payloadBytes = getPayload().getBytes();
```
Comment (ywang-nr) [inline agent-core/src/main/java/com/newrelic/agent/android/payload/PayloadSender.java:114]: get all the bytes might be a problem here

**Review by ywang-nr — COMMENTED** [review]
few comments with changes required

File: instrumentation/src/main/java/com/newrelic/agent/util/MapFileGenerator.java [hunk 4]
```diff
@@ -23,7 +23,7 @@
 
 public class MapFileGenerator {
 
-    private static final int KNOWN_MAP_SIZE = 91;
+    private static final int KNOWN_MAP_SIZE = 101;
```
[the remark this answers is not in the document]
Comment (ywang-nr) [reply instrumentation/src/main/java/com/newrelic/agent/util/MapFileGenerator.java:27]: all the jetpack are included, take them out and update this count

File: samples/agent-test-app/src/main/java/com/newrelic/agent/android/testapp/MainActivity.java [hunk 5]
```diff
@@ -63,6 +63,8 @@ protected void onCreate(Bundle savedInstanceState) {
 
         // Add NewRelic
         NewRelic.enableFeature(FeatureFlag.NativeReporting);
+        NewRelic.enableFeature(FeatureFlag.LogReporting);
```
Comment (ywang-nr) [inline samples/agent-test-app/src/main/java/com/newrelic/agent/android/testapp/MainActivity.java:66]: @cthomas-newrelic enable feature flag here

File: agent-core/build.gradle (hunk 1 above)
Comment (cthomas-newrelic) [reply agent-core/build.gradle:41]: It is necessary

File: agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestConfiguration.java [hunk 6]
```diff
@@ -274,7 +307,7 @@ public void setApplication_id(String application_id) {
 
     public String getAccount_id() {
         if (account_id == null) {
```
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestConfiguration.java:309]: It is not longer used in any important way, but will leave it in and refactor it out in later releases.

File: instrumentation/src/main/java/com/newrelic/agent/util/MapFileGenerator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply instrumentation/src/main/java/com/newrelic/agent/util/MapFileGenerator.java:27]: Leaving it in place, as a canary, so Jetpack is not included in release notes.

File: agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+
+        do {
+            closedLogFile = new File(LogReporter.logDataStore,
+                    String.format(Locale.getDefault(),
+                            LogReporter.LOG_FILE_MASK,
+                            System.currentTimeMillis(),
+                            LogReportState.CLOSED.extension));
+
+        } while (closedLogFile.exists() && 0 < closedLogFile.length());
+
```
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java:474]: Done in incoming PR.

File: agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java (hunk 2 above)
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/logging/LogReporter.java:71]: Done in incoming PR.

File: agent-core/src/main/java/com/newrelic/agent/android/payload/PayloadSender.java (hunk 3 above)
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/payload/PayloadSender.java:114]: Might be a problem in low memory conditions, we'll have to see. I will look for a more resilient solution.

File: samples/agent-test-app/src/main/java/com/newrelic/agent/android/testapp/MainActivity.java (hunk 5 above)
Comment (cthomas-newrelic) [reply samples/agent-test-app/src/main/java/com/newrelic/agent/android/testapp/MainActivity.java:66]: Done in incoming PR.

File: agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java [hunk 8]
```diff
@@ -145,20 +151,28 @@ public void shutdown() {
...
+     *
+     * @param bWait If true, wait for harvest completion
+     */
+    public void tickNow(boolean bWait) {
         try {
-            future.get();
+            // throttle on abuse
+//            if (timeSinceLastTick() <= TimeUnit.SECONDS.toMillis(15)) {
+//                log.warn("HarvestTimer.tickNow() called too frequently");
+//            } else {
```
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java:165]: done in incoming PR.

File: agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java [hunk 9]
```diff
@@ -145,20 +151,28 @@ public void shutdown() {
...
+            final HarvestTimer timer = this;
+            ScheduledFuture<?> future = scheduler.schedule(() -> timer.tick(), 0, TimeUnit.MILLISECONDS);
+            if (bWait && !future.isCancelled()) {
+                future.get();
+                startTimeMs = System.currentTimeMillis();
+            }
+
+//            }
         } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java:174]: Done in incoming PR.

File: agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java [hunk 10]
```diff
@@ -145,20 +151,28 @@ public void shutdown() {
...
-            future.get();
+            // throttle on abuse
+//            if (timeSinceLastTick() <= TimeUnit.SECONDS.toMillis(15)) {
+//                log.warn("HarvestTimer.tickNow() called too frequently");
+//            } else {
+            final HarvestTimer timer = this;
+            ScheduledFuture<?> future = scheduler.schedule(() -> timer.tick(), 0, TimeUnit.MILLISECONDS);
+            if (bWait && !future.isCancelled()) {
+                future.get();
+                startTimeMs = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (cthomas-newrelic) [reply agent-core/src/main/java/com/newrelic/agent/android/harvest/HarvestTimer.java:170]: Doen in incoming PR

**Review by cthomas-newrelic — DISMISSED** [review]
With merge of release/cst-PRRMods

**Review by JustinRush-NR — APPROVED** [review]
Good to merge

**Review by mbruin-NR — APPROVED** [review]

**Review by cdillard-NewRelic — COMMENTED** [review]
Approve

**Pull request merged**